src/bindings/python/flux/job/executor.py

```diff
@@ -0,0 +1,524 @@
+                daemon=True,
+            )
+            for i, deque in enumerate(self._submission_queues)
+        ]
+        for thread in self._executor_threads:
+            thread.start()
+        # register a finalizer to ensure worker threads are notified to shut down
+        self._finalizer = weakref.finalize(
+            self, self._shutdown_threads, self._shutdown_event, self._executor_threads
+        )
+
+    def shutdown(self, wait=True, *, cancel_futures=False):
+        """Clean-up the resources associated with the Executor.
+
+        It is safe to call this method several times. Otherwise, no other
+        methods can be called after this one.
+
+        :param wait: If ``True``, then shutdown will not return until all running
+            futures have finished executing and the resources used by the
+            executor have been reclaimed.
+        :param cancel_futures: If ``True``, this method will cancel all pending
+            futures that the executor has not started running. Any futures that
+            are completed or running won't be cancelled, regardless of the value
+            of ``cancel_futures``.
+        """
+        with self._shutdown_lock:
+            self._shutdown_event.set()
+        if cancel_futures:
+            # Drain all work items from the queues, and then cancel their
+            # associated futures.
+            for deque in self._submission_queues:
+                while deque:
+                    try:
+                        _, user_future = deque.popleft()
+                    except IndexError:
+                        pass
+                    else:
+                        user_future.cancel()
+                        user_future.set_running_or_notify_cancelled()
+        if wait:
+            for thread in self._executor_threads:
+                thread.join()
+
+    def submit(self, jobspec):
+        """Submit a jobspec to Flux and return a ``FluxExecutorFuture``."""
+        with self._shutdown_lock:
+            if self._shutdown_event.is_set():
+                raise RuntimeError("cannot schedule new futures after shutdown")
+            future_owner_id = self._executor_threads[self._next_thread].ident
+            fut = self._future_type(future_owner_id)
+            self._submission_queues[self._next_thread].append((jobspec, fut))
+            self._next_thread = (self._next_thread + 1) % len(self._submission_queues)
+            return fut
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.shutdown(wait=True)
+        return False
+
+    @staticmethod
+    def _shutdown_threads(event, threads):
+        """Set the threading.Event and join all threads.
+
+        Not a method so as not to prevent garbage collection
+        (see `weakref.finalize` docs).
+        """
+        event.set()
+        for thread in threads:
+            thread.join()
+
+
+class FluxEventExecutor(FluxExecutor):
```

Comment:
I think that confusion could be helped a lot by sphinx documentation, but yeah I am on the fence about `FluxExecutor`. I ran a handful of tests and found that `FluxEventExecutor` took 5-30% longer to complete the same number of jobs. I'll run more when I get the chance, since 5% is trivial but ~30% is somewhat significant (?)... I would like to know what other people think.

Reply:
A couple thoughts

 * could any overhead of watching individual job events be mitigated by only starting to watch events when at least one callback is installed? That is, you get the low overhead of `FluxExecutor` until you install the first callback for an event, and then you get the behavior of `FluxEventExecutor`?
 * I'm surprised that `FluxEventExecutor` takes longer to complete the same number of jobs, since I assume other parts of Flux limit the maximum job throughput, not the ability to get event notifications in a front-end script. I'd be interested in trying out your test as well.

Reply:
Later today or tomorrow I will provide the testing scripts and my own results. The results I mentioned above are from around two weeks ago and I didn't run many tests, so the results probably aren't all that reliable.

>     * could any overhead of watching individual job events be mitigated by only starting to watch events when at least one callback is installed? That is, you get the low overhead of `FluxExecutor` until you install the first callback for an event, and then you get the behavior of `FluxEventExecutor`?

I do like that idea but I don't know how to implement it well. Callbacks go on to `Future` objects, which don't do very much---they just hold data and locks for executors to use. You would need some way to alert the executor threads that event callbacks had been added, so they could request event updates. Probably this would have to be implemented as some kind of polling mechanism... but then there is the issue of timing. Even if users added event callbacks as soon as they got the future, there is no guarantee that by that time the executor won't have already completed and discarded the future.

What about instead an `event_updates=True` option in the `FluxExecutor` constructor? That would be pretty easy to implement.

Reply:
> What about instead an event_updates=True option in the FluxExecutor constructor? That would be pretty easy to implement. If True (the default) submit returns a future that receives event updates, otherwise it returns a future that doesn't even have the add_event_callback method.

￼Seems like a great idea to me! However, you should take @SteVwonder's feedback more seriously than mine on these Python interface questions.

Thanks!

Reply:
> Even if users added event callbacks as soon as they got the future, there is no guarantee that by that time the executor won't have already completed and discarded the future.

Yeah, I was worried about the same race condition

> What about instead an event_updates=True option in the FluxExecutor constructor? 

I like your suggestion.  The constructor returning different futures based on a keyword argument is a bit weird at first glance, but since the default will be to have all of the functionality, I think it is fine.  Anyone that "has the need for speed" can set the non-default value and hopefully know what the trade-offs are in doing so.  As you point out, readthedocs documentation will help there.

Speaking of RTD documentation, thanks for the excellent docstrings!

Reply:
I ran some performance tests. On one node of quartz, and using my `flux-core/src/cmd/flux` (I didn't build flux-sched, if that makes a difference), `FluxExecutor` was around 7% faster than `FluxEventExecutor` at running single-core `/usr/bin/true` jobs, and meanwhile the usual `bulksubmit.py` was 2% faster than `FluxExecutor`. I think those numbers are so low it doesn't make sense to keep `FluxExecutor` or even add an `event_updates` option to the constructor. If you two approve I'll rename `FluxEventExecutor` to `FluxExecutor` and remove the original `FluxExecutor`. 

My testing setup is available [here](https://github.com/jameshcorbett/flux-executor-tests) and here is a quick summary of my data followed by the full set:

```
                           jobs_per_second
implementation                            
bulksubmit                       21.301160
bulksubmit_event_executor        19.532555
bulksubmit_executor              20.839853
```


```
   hostname             implementation  jobcount            timestamp       timing
0    quartz                 bulksubmit      1000  2021-02-03 21:33:14    42.085684
1    quartz                 bulksubmit      1000  2021-02-03 21:35:36    39.843733
2    quartz                 bulksubmit      1000  2021-02-03 21:37:15    39.610015
3    quartz                 bulksubmit      1000  2021-02-03 21:39:43    41.635208
4    quartz                 bulksubmit      1000  2021-02-03 21:42:06    39.437209
5    quartz  bulksubmit_event_executor      1000  2021-02-03 21:34:50    43.579095
6    quartz  bulksubmit_event_executor      1000  2021-02-03 21:38:56    44.146753
7    quartz  bulksubmit_event_executor      1000  2021-02-03 21:41:21    46.492944
8    quartz  bulksubmit_event_executor      1000  2021-02-03 21:43:43    44.794582
9    quartz        bulksubmit_executor      1000  2021-02-03 21:34:02    42.607343
10   quartz        bulksubmit_executor      1000  2021-02-03 21:38:05    43.609146
11   quartz        bulksubmit_executor      1000  2021-02-03 21:40:29    40.705514
12   quartz        bulksubmit_executor      1000  2021-02-03 21:42:52    41.906976
13   quartz                 bulksubmit      2000  2021-02-03 21:45:16    88.055949
14   quartz                 bulksubmit      2000  2021-02-03 21:50:18    90.384738
15   quartz                 bulksubmit      2000  2021-02-03 21:55:08    89.209853
16   quartz  bulksubmit_event_executor      2000  2021-02-03 21:48:41    99.885325
17   quartz  bulksubmit_event_executor      2000  2021-02-03 21:53:32    94.196928
18   quartz  bulksubmit_event_executor      2000  2021-02-03 21:58:26    93.821270
19   quartz        bulksubmit_executor      2000  2021-02-03 21:46:55    92.382404
20   quartz        bulksubmit_executor      2000  2021-02-03 21:51:52    87.318732
21   quartz        bulksubmit_executor      2000  2021-02-03 21:56:46    90.404887
22   quartz                 bulksubmit      4000  2021-02-03 22:01:41   187.820985
23   quartz                 bulksubmit      4000  2021-02-03 22:11:50   194.037138
24   quartz                 bulksubmit      4000  2021-02-03 22:22:10   201.357420
25   quartz  bulksubmit_event_executor      4000  2021-02-03 22:08:30   201.805478
26   quartz  bulksubmit_event_executor      4000  2021-02-03 22:18:36   202.817005
27   quartz  bulksubmit_event_executor      4000  2021-02-03 22:29:18   216.205930
28   quartz        bulksubmit_executor      4000  2021-02-03 22:05:01   193.713691
29   quartz        bulksubmit_executor      4000  2021-02-03 22:15:07   189.601299
30   quartz        bulksubmit_executor      4000  2021-02-03 22:25:35   192.594310
31   quartz                 bulksubmit      8000  2021-02-03 22:36:28   419.229630
32   quartz                 bulksubmit      8000  2021-02-03 22:58:40   424.271110
33   quartz                 bulksubmit      8000  2021-02-03 23:20:50   424.680332
34   quartz  bulksubmit_event_executor      8000  2021-02-03 22:51:28   459.929019
35   quartz  bulksubmit_event_executor      8000  2021-02-03 23:13:36   461.522901
36   quartz        bulksubmit_executor      8000  2021-02-03 22:43:41   424.084636
37   quartz        bulksubmit_executor      8000  2021-02-03 23:05:46   419.462058
38   quartz        bulksubmit_executor      8000  2021-02-03 23:28:10   433.009854
39   quartz                 bulksubmit     16000  2021-02-04 00:01:34  1004.535738
40   quartz                 bulksubmit     16000  2021-02-04 00:54:34  1025.751971
41   quartz  bulksubmit_event_executor     16000  2021-02-04 00:37:19  1127.634311
42   quartz  bulksubmit_event_executor     16000  2021-02-04 01:29:44  1104.115855
43   quartz        bulksubmit_executor     16000  2021-02-04 00:18:23   999.608337
44   quartz        bulksubmit_executor     16000  2021-02-04 01:11:12   990.196928
```

Reply:
Oh yeah, and there is also the `waitable` thing going against `FluxExecutor` and in favor of `FluxEventExecutor`, since `FluxExecutor` requires waitable jobs and `FluxEventExecutor` doesn't. But `FluxExecutor` does guarantee that every `waitable` job is in fact waited on, so I'm not sure how much of an advantage that is.

Reply:
I think your benchmarks are hitting the limit of how fast Flux can run *actual* jobs on quartz, and not really testing any difference between the ability of `FluxExecutor` and `FluxEventExecutor` to handle the events that indicate a job is complete. The method via which jobs are submitted should not influence the rate at which jobs run and complete (except if your job submission rate was slower that the throughput rate on the system you are testing, not at all the case with asynchronous submission)

The real job throughput is currently limited by the job execution system, so it might be interesting to include tests which use simulated execution. The job-exec module will skip actual execution and only simulate it if the key `attributes.system.exec.test.run_duration` is set in the submitted jobspec, e.g. with the following change to your tests:

```diffdiff --git a/bulksubmit_executor.py b/bulksubmit_executor.py
index e8b4c26..f7a6518 100755
--- a/bulksubmit_executor.py
+++ b/bulksubmit_executor.py
@@ -40,6 +40,7 @@ def main():
     exec_type = FluxExecutor if not args.events else FluxEventExecutor
     with exec_type() as executor:
         compute_jobspec = JobspecV1.from_command(args.command)
+        compute_jobspec.setattr("system.exec.test.run_duration", "0.001s")
         futures = [executor.submit(compute_jobspec) for _ in range(args.njobs)]
         # wait for jobs to complete
         done, not_done = cf.wait(futures, return_when=cf.FIRST_EXCEPTION)
```

jobs will be executed with a simulated runtime of 1ms (essentially nothing).

This will get the actual job throughput a lot higher and may show more difference between `FluxExecutor` and `FluxEventExecutor` *if* watching events happens to cause the event executor to "lag behind" actual job execution. (I do kind of doubt this is this case though, since the max job throughput even with simulated execution is at most 150job/s)

Oh, also either rebase this PR on latest master or reload sched-simple with "unlimited" option to get the highest possible throughput.

Reply:
BTW, one way to measure the job throughput independent of watching events or job wait() performance is to read the eventlog timestamp for the `submit` event of the first submitted job, and the timestamp of the `clean` event for the last job. 

This can be used to compare how the long the script took vs the actual makespan of the simulated workload. We'd be looking for a large difference if events can't be processed as quickly as they are generated, but I don't (currently) see this on my test system:

```
ƒ(s=8,d=0,builddir) grondo@asp:~/git/f.git$ ./throughput.py -n 1000
number of jobs: 1000
script runtime: 7.595 s
job runtime:    7.364 s
throughput:     135.8 job/s (script: 131.7 job/s)
```

Reply:
I implemented your first suggestion (rebase with master and add the `attributes.system.exec.test.run_duration` line) and here are my summary results:

```
                           jobs_per_second
implementation                            
bulksubmit                       97.274905
bulksubmit_event_executor        79.491056
bulksubmit_executor              96.978769
```

A bigger difference. So something about the `FluxEventExecutor` is definitely slower. Based on my original tests, though, it seems that the performance of `FluxEventExecutor` is good enough for any real workflow. But if Flux's job's throughput rose substantially, so that it could complete >100 jobs/sec, and there was an actual use-case that wanted >100 jobs/sec, then `FluxExecutor` might be better.

Reply:
I removed ``FluxExecutor`` and replaced it with ``FluxEventExecutor``. Once a real workflow is being limited by ``FluxEventExecutor``, then (and only then) I think it makes sense to offer something else that is faster.